src/sco.c

```diff
@@ -208,66 +217,118 @@ int sco_setup_connection_dispatcher(struct ba_adapter *a) {
 	return 0;
 }
 
+static enum ba_transport_thread_signal sco_enc_io_poll_signal_filter(
+		enum ba_transport_thread_signal signal, void *userdata) {
+	struct sco_transfer *transfer = userdata;
+	if (signal == BA_TRANSPORT_THREAD_SIGNAL_BT_MTU) {
+		debug("Setting SCO write MTU %zd", transfer->transport->mtu_write);
+		transfer->bytes = transfer->transport->mtu_write;
+		transfer->samples = transfer->bytes / sizeof(uint16_t);
+		size_t bufsize = transfer->samples * 4;
+		if (transfer->buffer != NULL && transfer->buffer->nmemb < bufsize)
+			ffb_init_int16_t(transfer->buffer, bufsize);
+	}
+	return signal;
+}
+
+static void	sco_update_mtu(struct ba_transport *t, size_t len) {
+	t->mtu_write = len;
+	ba_transport_thread_signal_send(&t->thread_enc, BA_TRANSPORT_THREAD_SIGNAL_BT_MTU);
+}
+
 static void *sco_cvsd_enc_thread(struct ba_transport_thread *th) {
 
 	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
 	pthread_cleanup_push(PTHREAD_CLEANUP(ba_transport_thread_cleanup), th);
 
 	struct ba_transport *t = th->t;
 	struct ba_transport_pcm *pcm = &t->sco.spk_pcm;
-	struct io_poll io = { .timeout = -1 };
-
-	const size_t mtu_samples = t->mtu_write / sizeof(int16_t);
-	const size_t mtu_write = t->mtu_write;
 
 	ffb_t buffer = { 0 };
+
+	struct sco_transfer transfer = {
+		.bytes = 0,
+		.samples = 0,
+		.buffer = &buffer,
+		.transport = t
+	};
+
+	struct io_poll io = {
+		.signal.filter = sco_enc_io_poll_signal_filter,
+		.signal.userdata = &transfer,
+		.timeout = -1,
+	};
+
 	pthread_cleanup_push(PTHREAD_CLEANUP(ffb_free), &buffer);
 
-	/* define a bigger buffer to enhance read performance */
-	if (ffb_init_int16_t(&buffer, mtu_samples * 4) == -1) {
+	/* The buffer will be resized once the SCO transfer size is known. We
+	 * choose a small value initially (10ms) to ensure that we discard as little
+	 * audio as possible. */
+	if (ffb_init_int16_t(&buffer, 80) == -1) {
 		error("Couldn't create data buffer: %s", strerror(errno));
 		goto fail_init;
 	}
 
 	debug_transport_thread_loop(th, "START");
-	for (ba_transport_thread_set_state_running(th);;) {
+	ba_transport_thread_set_state_running(th);
 
-		ssize_t samples = ffb_len_in(&buffer);
-		if ((samples = io_poll_and_read_pcm(&io, pcm, buffer.tail, samples)) <= 0) {
+	ssize_t samples;
+	/* discard samples until SCO transfer size is known. */
+	while (transfer.bytes == 0) {
```

Comment:
What will happen if there will be no incoming SCO stream? I now that such a case will not happen (or will it?) in real life, but in principle one can use two blualsa-based devices and try to transfer SCO data only in one direction (like with A2DP).

Reply:
I've tested bluealsa-bluealsa, and can confirm that a SCO transport *always* delivers packets in both directions between the adapters as soon as the transport is acquired, whether or not any valid audio data is supplied over the HCI. The only exception, I think, is if the adapter does not route incoming packets to the HCI, in which case bluealsa will not see them. I have included a timeout in the decoder thread to catch this case.